### PR TITLE
Add Teads Adapter

### PR DESCRIPTION
**Pull request by AntoxaAntoxic**

### CODE REVIEW

File: src/main/resources/static/bidder-params/teads.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+  "type": "object",
+  "properties": {
+    "placementId": {
+      "type": "integer",
+      "description": "The placement id.",
+      "minimum": 1
+    }
+  },
+  "required": ["placementId"]
```
Comment (SerhiiNahornyi) [inline src/main/resources/static/bidder-params/teads.json:14]: Please format with default formatter

File: src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+                .build();
+    }
+
+    private static Banner modifyBanner(Banner banner) {
+        return banner != null
+                ? Stream.ofNullable(banner.getFormat())
+                .flatMap(Collection::stream)
+                .findFirst()
+                .map(format -> banner.toBuilder().w(format.getW()).h(format.getH()).build())
+                .orElse(banner)
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:86]: Let's use more common approaches for this logic, like in GammaBidder

File: src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    public TeadsBidder(String endpointUrl, JacksonMapper mapper) {
+        this.endpointUrl = HttpUtil.validateUrl(Objects.requireNonNull(endpointUrl));
+        this.mapper = Objects.requireNonNull(mapper);
+    }
+
+    @Override
+    public Result<List<HttpRequest<BidRequest>>> makeHttpRequests(BidRequest request) {
+        try {
+            final List<Imp> modifiedImps = request.getImp().stream().map(this::modifyImp).collect(Collectors.toList());
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:58]: Please separate chain calls with \n

File: src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    private TeadsImpExt parseImpExt(Imp imp) {
+        try {
+            return mapper.mapper().convertValue(imp.getExt(), TYPE_REFERENCE).getBidder();
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage());
+        }
+    }
+
+    private HttpRequest<BidRequest> makeHttpRequest(BidRequest bidRequest) {
+
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:99]: No need for \n

File: src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        final ExtBidPrebidMeta meta = parseExtBidPrebidMeta(bid);
+        if (StringUtils.isBlank(meta.getRendererName())) {
+            throw new PreBidException("RendererName should not be empty");
+        }
+
+        if (StringUtils.isBlank(meta.getRendererVersion())) {
+            throw new PreBidException("RendererVersion should not be empty");
+        }
+
+        return BidderBid.of(bid, bidType, currency);
```
Comment (SerhiiNahornyi) [inline src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:151]: Looks like In Go they completely creates new meta-object, with just two fields

**Review by SerhiiNahornyi — CHANGES_REQUESTED** [review]

File: src/main/resources/static/bidder-params/teads.json (hunk 1 above)
Comment (AntoxaAntoxic) [reply src/main/resources/static/bidder-params/teads.json:14]: it's already formatted

File: src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java (hunk 5 above)
Comment (AntoxaAntoxic) [reply src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:151]: that's not true due to the test cases they cover, please check some of them

Comment (SerhiiNahornyi) [reply src/main/java/org/prebid/server/bidder/teads/TeadsBidder.java:153]: // TypedBid.BidMeta will become "response.seatbid[i].bid.ext.prebid.meta" in the final OpenRTB response

They have this data only in bid, but in final version they will have only this two fields

**Review by SerhiiNahornyi — APPROVED** [review]

**Pull request merged**